[Feat] 로그인화면의 UI를 구현했습니다.

snughnu opened the pull request:
# 배경

<img width="278" alt="스크린샷 2024-11-20 01 01 47" src="https://github.com/user-attachments/assets/f2c08f39-f758-4276-9e38-fc42488dee39">

#9, #11  로그인 화면의 UI를 구현합니다.

# 작업 내용

https://github.com/user-attachments/assets/63204957-eb2a-4097-8c0e-3d70788bbddd

## UI구현
- 화면에 보이는 이미지, 텍스트필드, 버튼, 구분(소셜로그인)선, 소셜로그인을 구현했습니다.
- 유저가 구분하기 쉽게 선택된 텍스트필드는 색상의 변화를 주었습니다.
- 유저가 불편함 없이 다음 동작으로 넘어가기 위해 텍스트필드에서 return버튼을 눌렀을 때의 동작을 구현했습니다.
- 불필요한 서버 통신을 막기 위해 이메일과 비밀번호가 모두 입력되었을 때 로그인 버튼이 활성화 되도록 구현했습니다.
- 소셜로그인 부분은 가이드라인에 맞추기 위해 조사 이후 작업 예정입니다.

# 테스트 방법
- 프리뷰 코드를 사용해서 UI를 확인할 수 있습니다.

# 리뷰 노트
- 소셜로그인 부분은 정해진 가이드라인을 공부 후 작업하겠습니다.
- 로그인 버튼의 활성화 코드는 View에 집중된 작업이라 생각해서 LoginView.swift에 작업했습니다.
- 커밋 컨벤션에 맞게 작성하도록 하겠습니다.

snughnu committed:
feat: Common LoginTextField 구현

snughnu committed:
feat/Common-LoginTextField 수정

snughnu committed:
feat/Login UI구현

snughnu committed:
feat/Login UI 수정

snughnu committed:
feat/로그인버튼 활성화

snughnu committed:
feat/텍스트필드기능 - 텍스트필드 이외의 곳 터치

snughnu committed:
feat/텍스트필드-프로토콜-텍스트필드 색 변화 구현

snughnu committed:
feat/Controller base코드 작성

snughnu committed:
feat/텍스트필드 - return 버튼 기능 추가

snughnu committed:
feat/텍스트필드 프로토콜 Controller로 이동

MinwooJe commented on Popcorn-iOS/Source/Presentation/Login/Common/View/LoginTextField.swift:
```diff
@@ -0,0 +1,51 @@
+import UIKit
+
+final class LoginTextField: UITextField {
+    private let insets: UIEdgeInsets = .init(top: 16, left: 16, bottom: 16, right: 40)
```
p4) 가독성 차원에서생성자랑 위의 코드랑 공백이 있음 좋을 것 같습니다~

snughnu replied:
넵!

MinwooJe commented on Popcorn-iOS/Source/Presentation/Login/Common/View/LoginTextField.swift:
```diff
@@ -0,0 +1,51 @@
+        return bounds.inset(by: insets)
+    }
+}
+
```
p4) extension 위에는 `// Mark: - ` 를 이용하는건 어떤가요?이 부분은 컨벤션으로 정하진 않았네요ㅠ

snughnu replied:
앞으로 extension위에 `// Mark: -`을 이용하는걸 기본으로 할게요!

MinwooJe commented on Popcorn-iOS/Source/Presentation/Login/Common/View/LoginTextField.swift:
```diff
@@ -0,0 +1,51 @@
+import UIKit
+
+final class LoginTextField: UITextField {
+    private let insets: UIEdgeInsets = .init(top: 16, left: 16, bottom: 16, right: 40)
```
p3)
<img width="352" alt="스크린샷 2024-11-20 오후 2 19 14" src="https://github.com/user-attachments/assets/f3597ebd-48bc-49c8-a628-ebaea84143a2">

- UIEdgeInset의 `left`가 디자인이랑 다릅니다!
- 아이디가 엄청 길 경우를 고려해서 `right`의 값을 `left`를 기준으로 맞추는건 어떤가요?

snughnu replied:
레이아웃이 확정나지 않은거 같아서 임의로 작성해놨습니다!
우선은 아이디가 엄청 길 경우를 고려하는게 맞으니 좌우 20으로 맞추겠습니다.

MinwooJe replied:
넵!

MinwooJe reviewed (commented):
- 다음부터 Asset을 넣는 경우 따로 커밋을 분리하면 좋을 것 같습니다!
- 커밋 메시지가 텍스트필드 수정보다는 색상 수정이라면 리뷰어가 가벼운 마음으로 넘어갈 수 있을 것 같아서 조금 더 상세한 메시지를 적어주시면 좋을 것 같습니다~~

MinwooJe commented on Popcorn-iOS/Source/Presentation/Login/LoginScene/View/LoginView.swift:
```diff
@@ -7,14 +7,272 @@
+        return button
+    }()
+
+    lazy var emailPasswordLoginStackView: UIStackView = {
```
p3)
이메일 텍스트필드와, 비밀번호 텍스트 필드의 간격은 20이고, 비밀번화 텍스트필드와 로그인 버튼의 간격은 27입니다!
스택뷰를 하나 더 만들거나 오토레이아웃 제약을 거는 방식으로 변경해야 할 것 같습니다~

snughnu replied:
레이아웃이 확정나지 않은거 같아서 임의로 작성해놨습니다.
회의때 디자인이 확정인지 여쭤보고 수정하겠습니다!

MinwooJe replied:
넵!

MinwooJe commented on Popcorn-iOS/Source/Presentation/Login/LoginScene/View/LoginView.swift:
```diff
@@ -7,14 +7,272 @@
+        button.backgroundColor = .clear
+        button.titleLabel?.textAlignment = .center
+        button.titleLabel?.adjustsFontSizeToFitWidth = true
+        button.titleLabel?.minimumScaleFactor = 1.0
```
p2)
minimumScaleFactor를 1.0으로 준 이유가 있을까요?
제가 저 프로퍼티를 안써봐서 잘은 모르지만, 1.0으로 설정하면 결국엔 글자가 안줄어드는것과 동일하니 디폴트 값인 0과 똑같은 효과가 날 것 같다고 생각되네요..

알려주시면 감사하겠습니다!

snughnu replied:
0은 100%, 0.5는 50%, 1.0은 0% 줄어들 수 있습니다.
텍스트 크기를 설정해놨으니 줄어들지 않도록 1.0의 값을 사용했습니다!

MinwooJe replied:
레이블의 텍스트 크기는 기본적으로 줄어들지 않습니다!
아마 `adjustsFontSizeToFitWidth` 설정이 되어있어서 줄어들지 않았나 싶습니다~

https://developer.apple.com/documentation/uikit/uilabel/1620544-minimumscalefactor
공식문서를 참고하시면 `minimumScaleFactor`의 디폴트는 0이고, 0일 때는 텍스트 크기를 축소하지 않는다고 되어있습니다!

MinwooJe commented on Popcorn-iOS/Source/Presentation/Login/LoginScene/View/LoginView.swift:
```diff
@@ -7,14 +7,272 @@
+    }
+}
+
+// MARK: - configure
```
P5) 무엇을 configure 하나요?ㅎ

snughnu replied:
layout 입니다. 수정할게요~

MinwooJe commented on Popcorn-iOS/Source/Presentation/Login/LoginScene/View/LoginView.swift:
```diff
@@ -7,14 +7,272 @@
+        $0.translatesAutoresizingMaskIntoConstraints = false
+        }
+    }
+    private func configureLayout() {
```
P2) 레이아웃을 잡는 함수가 불필요하게 많은 것 같습니다!
`configureLayout()` 메서드에 모든 레이아웃 코드들을 넣어도 괜찮지 않을까요?
디버깅하는데도 큰 문제가 없다고 생각됩니다..!

snughnu replied:
레이아웃잡을 때 의도대로 되지 않으면 `configureLayout`함수에서 파악하기 좋아서 구분해봤습니다.

MinwooJe commented on Popcorn-iOS/Source/Presentation/Login/LoginScene/View/LoginView.swift:
```diff
@@ -7,14 +7,272 @@
+
+    private let rightSeparateView: UIView = {
+        let view = UIView()
+        view.heightAnchor.constraint(equalToConstant: 1).isActive = true
```
p3) 뷰의 제약을 잡는 코드를 여기에 작성한 이유가 있을까요?
아래 `configureLayout()` 메서드안에 넣는게 낫지 않을까요?

snughnu replied:
구분선의 경우는 변경되지 않을거같아서
'뷰를 올려서 높이를 1로한다'가 아닌 '높이가 1인 뷰를 올린다'로 생각해서 작성했습니다.

MinwooJe commented on Popcorn-iOS/Source/Presentation/Login/LoginScene/View/LoginView.swift:
```diff
@@ -7,14 +7,272 @@
+
+    private let findSignUpSeparateView: UIView = {
+        let view = UIView()
+        view.widthAnchor.constraint(equalToConstant: 1).isActive = true
```
p3) 뷰의 제약을 잡는 코드를 여기에 작성한 이유가 있을까요?
아래 configureLayout() 메서드안에 넣는게 낫지 않을까요?

snughnu replied:
구분선의 경우는 변경되지 않을거같아서
'뷰를 올려서 너비를 1로한다'가 아닌 '너비가 1인 뷰를 올린다'로 생각해서 작성했습니다.

MinwooJe commented on Popcorn-iOS/Source/Presentation/Login/LoginScene/View/LoginView.swift:
```diff
@@ -7,14 +7,272 @@
 
-class LoginView: UIView {
+final class LoginView: UIView {
+    private var popcornImageView: UIImageView = {
```
p3) 팝콘 이미지와 팝콘 텍스트 이미지가 같이 있는게 나을 것 같습니다!
이런 부분은 디자이너님께 하나의 레이어? 프레임?으로 합쳐달라고 한 뒤 작업하시는게 더 편할거에요!

snughnu replied:
넵! 디자인이 확정나면 수정하겠습니다.

MinwooJe commented on Popcorn-iOS/Source/Presentation/Login/LoginScene/View/LoginView.swift:
```diff
@@ -7,14 +7,272 @@
+        button.setTitle("로그인", for: .normal)
+        button.setTitleColor(.white, for: .normal)
+        button.backgroundColor = #colorLiteral(red: 1.0, green: 0.384, blue: 0.004, alpha: 1.0)
+        button.layer.cornerRadius = 10
```
p5) 제가 지금 작업중인 pr에 `cornerRadius` 편하게 잡는 익스텐션을 추가했습니다! PR 올라가면 확인하시고 수정해주세요~

snughnu replied:
넵! 헷갈리면 알아보고 여쭤보도록 하겠습니다.

MinwooJe commented on Popcorn-iOS/Source/Presentation/Login/LoginScene/View/LoginView.swift:
```diff
@@ -7,14 +7,272 @@
+        let button = UIButton()
+        button.setTitle("회원가입", for: .normal)
+        button.setTitleColor(.black, for: .normal)
+//        button.titleLabel?.font = .systemFont(ofSize: 17, weight: .medium)
```
p4) 불필요한 주석은 삭제해주세요

MinwooJe commented on Popcorn-iOS/Source/Presentation/Login/LoginScene/View/LoginView.swift:
```diff
@@ -7,14 +7,272 @@
+
+    private let leftSeparateView: UIView = {
+        let view = UIView()
+        view.heightAnchor.constraint(equalToConstant: 1).isActive = true
```
p3) 마찬가지입니다!

snughnu replied:
넵

MinwooJe commented on Popcorn-iOS/Source/Presentation/Login/LoginScene/View/LoginView.swift:
```diff
@@ -7,14 +7,272 @@
+        NSLayoutConstraint.activate([
+            popcornImageView.leadingAnchor.constraint(equalTo: safeAreaLayoutGuide.leadingAnchor, constant: 168),
+            popcornImageView.centerXAnchor.constraint(equalTo: safeAreaLayoutGuide.centerXAnchor),
+            popcornImageView.widthAnchor.constraint(equalToConstant: 58),
```
p2) width, height가 고정값으로 두면 다양한 화면 크기에 대한 대응이 힘들지 않을까요? 
상대적으로 제약을 잡는 오토레이아웃의 장점이 사라지는 것 같습니다
se, max, 16 기기별로 시뮬레이터를 돌려보시면 좋을 것 같애요!

snughnu replied:
팝콘 이미지를 피그마 디자인에 있는 제약값을 입력하면 다르게 나와서 고정해놨습니다
팝콘 이미지와 팝콘 텍스트가 합쳐지면 수정해겠습니다!

MinwooJe commented on Popcorn-iOS/Source/Presentation/Login/LoginScene/View/LoginView.swift:
```diff
@@ -7,14 +7,272 @@
+        NSLayoutConstraint.activate([
+            popcornLogoView.leadingAnchor.constraint(equalTo: safeAreaLayoutGuide.leadingAnchor, constant: 100),
+            popcornLogoView.centerXAnchor.constraint(equalTo: safeAreaLayoutGuide.centerXAnchor),
+            popcornLogoView.widthAnchor.constraint(equalToConstant: 122),
```
p2) 마찬가지입니다!

MinwooJe commented on Popcorn-iOS/Source/Presentation/Login/LoginScene/View/LoginView.swift:
```diff
@@ -35,13 +35,14 @@ final class LoginView: UIView {
         button.setTitle("로그인", for: .normal)
         button.setTitleColor(.white, for: .normal)
-        button.backgroundColor = #colorLiteral(red: 1.0, green: 0.384, blue: 0.004, alpha: 1.0)
+        button.backgroundColor = .lightGray
```
<img width="166" alt="스크린샷 2024-11-20 오후 2 52 26" src="https://github.com/user-attachments/assets/97b4a9bf-1473-4c2f-b778-56ebf2648feb">
p3) 이 색으로 통일시키는게 낫지 않을까요?

snughnu replied:
디자인에 없는 부분이라 임의로 설정했는데
다음 뷰의 버튼색과 통일하면 될거같네요!

MinwooJe commented on Popcorn-iOS/Source/Presentation/Login/LoginScene/View/LoginView.swift:
```diff
@@ -183,15 +182,23 @@ final class LoginView: UIView {
 
+// MARK: - setup
+extension LoginView {
+    func setup() {
```
p3) setup이라는 함수명이 어떤 동작을 하는지 불분명 한 것 같습니다
초기 세팅이라면 `configureInitialSetting` 같은 함수 명이 더 낫지 않을까요?
`configureSubviews`와 `configureLayout` 함수를 `setup`에 넣은 이유가 있을까요?

snughnu replied:
초기 세팅의 의미였습니다 `configureInitialSetting`로 수정할게요
`configureSubviews`와 `configureLayout`를 하는 것도 초기 세팅에 포함돼서 넣어놨습니다.

MinwooJe commented on Popcorn-iOS/Source/Presentation/Login/LoginScene/View/LoginView.swift:
```diff
@@ -75,7 +75,7 @@ final class LoginView: UIView {
         let view = UIView()
         view.widthAnchor.constraint(equalToConstant: 1).isActive = true
-        view.heightAnchor.constraint(equalToConstant: 20).isActive = true
+        view.heightAnchor.constraint(equalToConstant: 15).isActive = true
```
p3) 피그마 디자인이랑 높이가 다른 것 같습니다~

snughnu replied:
넵 수정했습니다

MinwooJe commented on Popcorn-iOS/Source/Presentation/Login/LoginScene/View/LoginView.swift:
```diff
@@ -196,6 +196,8 @@ extension LoginView {
         configureSubviews()
         configureLayout()
+        emailTextField.addTarget(self, action: #selector(textFieldEditingChanged(_:)), for: .editingChanged)
+        passwordTextField.addTarget(self, action: #selector(textFieldEditingChanged(_:)), for: .editingChanged)
```
p3) addTarget 하는 함수와 레이아웃 관련 함수는 분리시키는게 좋지 않을까요?

snughnu replied:
textField설정하는 함수를 분리하는게 좋겠네요!

snughnu committed:
refactor: 피드백내용 수정

- LoginTextField파일 생성자와 insets줄 공백 추가
- LoginTextField파일 extension위에 mark추가
- LoginTextField파일 insets수정
- LoginView파일 mark: configure -> mark: configure Layout으로 수정
- LoginView파일 LoginButton비활성 색상을 변경
- LoginView파일 setup함수명을 configureInitialSetting으로 변경
- LoginView파일 찾기/회원가입 사이즈 수정
- LoginView파일 setup(configureInitialSetting)함수에서 textField부분 분리

snughnu committed:
rename: 로그인화면의 이미지 삭제후 다시 삽입

snughnu committed:
design: 로그인뷰 로고 이미지 변경

snughnu committed:
design: 로그인뷰 레이아웃 수정 및 로고 이미지뷰 통합

snughnu committed:
refactor: 로그인뷰 오토레이아웃 코드구조 변경

snughnu committed:
style: 주석 대소문자 변경, configureInitialSetting함수 분리
